drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,645 @@
+					uart_irq_callback_t cb)
+{
+	__ASSERT(dev, "NULL pointer");
+
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	data->cb = cb;
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+
+static const struct uart_driver_api uart_nrfx_uarte_driver_api = {
+	.poll_in		= uarte_nrfx_poll_in,
+	.poll_out		= uarte_nrfx_poll_out,
+	.err_check		= uarte_nrfx_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill		= uarte_nrfx_fifo_fill,
+	.fifo_read		= uarte_nrfx_fifo_read,
+	.irq_tx_enable		= uarte_nrfx_irq_tx_enable,
+	.irq_tx_disable		= uarte_nrfx_irq_tx_disable,
+	.irq_tx_ready		= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_enable		= uarte_nrfx_irq_rx_enable,
+	.irq_rx_disable		= uarte_nrfx_irq_rx_disable,
+	.irq_tx_complete	= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_ready		= uarte_nrfx_irq_rx_ready,
+	.irq_err_enable		= uarte_nrfx_irq_err_enable,
+	.irq_err_disable	= uarte_nrfx_irq_err_disable,
+	.irq_is_pending		= uarte_nrfx_irq_is_pending,
+	.irq_update		= uarte_nrx_irq_update,
+	.irq_callback_set	= uarte_nrfx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uarte_instance_init(struct device *dev,
+			       const struct uarte_init_config *config,
+			       u8_t interrupts_active)
+{
+	__ASSERT(dev, "NULL pointer");
+
+	int err;
+	struct device *gpio_dev;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	__ASSERT(gpio_dev,
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	gpio_pin_write(gpio_dev, config->pseltxd, 1);
+	gpio_pin_configure(gpio_dev,
+			   config->pseltxd,
+			   GPIO_DIR_OUT);
+	gpio_pin_configure(gpio_dev,
+			   config->pselrxd,
+			   GPIO_DIR_IN);
+
+	nrf_uarte_txrx_pins_set(uarte,
+				config->pseltxd,
+				config->pselrxd);
+
+	if (config->hwfc == NRF_UARTE_HWFC_ENABLED) {
+		gpio_pin_write(gpio_dev, config->pselrts, 1);
+		gpio_pin_configure(gpio_dev,
+				   config->pselrts,
+				   GPIO_DIR_OUT);
+		gpio_pin_configure(gpio_dev,
+				   config->pselcts,
+				   GPIO_DIR_IN);
+		nrf_uarte_hwfc_pins_set(uarte,
+					config->pselrts,
+					config->pselcts);
+	}
+
+	/* Set low control and parity bit */
+	nrf_uarte_configure(uarte,
+			    config->parity,
+			    config->hwfc);
+
+	/* Set baud rate */
+	err = baudrate_set(dev, config->baudrate);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	nrf_uarte_enable(uarte);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	if (interrupts_active) {
+		u8_t dummy_byte = 0;
+
+		irq_enable(NRFX_IRQ_NUMBER_GET(uarte));
+
+		/* Set ENDTX event by sending empty frame */
+		nrf_uarte_tx_buffer_set(uarte, &dummy_byte, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
```

Comment:
To save this byte I've used `interrupts_active` variable.
I would like to avoid passing NULL pointer. According to spec:

"If the TXD.PTR and the RXD.PTR are not pointing to the Data RAM region, an EasyDMA transfer may result in a HardFault or RAM corruption."

Reply:
But `interrupts_active` is a local variable and the compiler may use a core register for storing it. By using a pointer to this variable you are forcing it to be stored on the stack. Thus, this way you are saving one byte in RAM but wasting a couple in flash.
I may be wrong, but the statement from the specification refers primarily to the situation when the transfer is actually performed, i.e. the pointed memory is accessed. But here we've got a fake transfer and no memory access is needed, so `NULL` should work.
Anyway, if you want to use a real Data RAM pointer here, use `data` or `&data->rx_data`. And add a proper comment, please, so no one wonders why such pointer is passed as the parameter.